### PR TITLE
Implement new dashboard API

### LINKED ISSUE #910: Adapt recent activities widget for mobile clients
### Is your feature request related to a problem? Please describe.

With https://github.com/nextcloud/server/pull/33658 mobile clients can access app widget data.

### Describe the solution you'd like

Adapt the *Upcoming events* widget to https://github.com/nextcloud/server/pull/33658 once it's done.

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by eneiluj**
- [x] requires https://github.com/nextcloud/server/pull/33658

closes #910 

Implement IAPIWIdget, IButtonWidget and IIconWidget.

### CODE REVIEW

File: lib/Dashboard/ActivityWidget.php [hunk 1]
```diff
@@ -83,6 +118,49 @@ public function getUrl(): ?string {
...
+			'',
+			0
+		);
+		return array_map(function (array $activity) {
+			return new WidgetItem(
+				$activity['subject'],
+				$this->dateTimeFormatter->formatTimeSpan($activity['timestamp']),
+				$activity['link'],
+				$this->urlGenerator->getAbsoluteURL(
+					$this->urlGenerator->linkToRoute('core.avatar.getAvatar', ['userId' => $activity['affecteduser'], 'size' => 44])
```
Comment (nickvergessen) [inline lib/Dashboard/ActivityWidget.php:145]: affecteduser is always "You". I don't think that is the icon we should show. Should use the `$event->getIcon()` instead?

File: lib/Dashboard/ActivityWidget.php [hunk 2]
```diff
@@ -26,22 +26,48 @@
...
 
 	/**
 	 * ActivityWidget constructor.
 	 * @param IL10N $l10n
+	 * @param Data $data
+	 * @param GroupHelper $helper
+	 * @param UserSettings $settings
+	 * @param IURLGenerator $urlGenerator
+	 * @param IDateTimeFormatter $dateTimeFormatter
 	 */
```
Comment (nickvergessen) [inline lib/Dashboard/ActivityWidget.php:58]:
```suggestion
```

**Review by nickvergessen — CHANGES_REQUESTED** [review]

File: lib/Dashboard/ActivityWidget.php [hunk 3]
```diff
@@ -83,6 +118,49 @@ public function getUrl(): ?string {
...
+	/**
+	 * @inheritDoc
+	 */
+	public function getItems(string $userId, ?string $since = null, int $limit = 7): array {
+		$activities = $this->data->get(
+			$this->helper,
+			$this->settings,
+			$userId,
+			$since ? (int) $since : 0,
+			50,
```
[the remark this answers is not in the document]
Comment (eneiluj) [reply lib/Dashboard/ActivityWidget.php:133]: Yes in the end we want 7 widget items. As I don't want to process the result of `data->get()` to "ungroup" the activities, I need to ask for more than 7 activity entries to make sure I get at least what I want. If there's a better way, I'm very open to it.

File: lib/Dashboard/ActivityWidget.php (hunk 1 above)
Comment (eneiluj) [reply lib/Dashboard/ActivityWidget.php:145]: Yeah I meant to use the `user` prop just like the frontend widget does. Anyway, you're right, the `icon` is better (but not consistent with the "javascript" widget).

File: lib/Dashboard/ActivityWidget.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (nickvergessen) [reply lib/Dashboard/ActivityWidget.php:124]: Yeah as said, no issue with that. just leave a comment in the code explaining it :)

File: lib/Dashboard/ActivityWidget.php (hunk 1 above)
Comment (nickvergessen) [reply lib/Dashboard/ActivityWidget.php]: The webfrontend shows the avatar + activity icon
https://github.com/nextcloud/activity/pull/920

Comment (eneiluj) [reply lib/Dashboard/ActivityWidget.php]: Ok then the action icon might be better than the user avatar. Done.

File: lib/Dashboard/ActivityWidget.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (eneiluj) [reply lib/Dashboard/ActivityWidget.php:126]: Done

**Comment by PVince81** [comment]
@eneiluj can you make CI happy ?

**Comment by nickvergessen** [comment]
CI will be happy after Server PR is in and then the CI dependency is bumped (happens on the weekend automatically, but can also be done manually on the day after server was merged): https://github.com/nextcloud/activity/blob/master/composer.json#L9

**Review by PVince81 — APPROVED** [review]
:+1:

File: lib/Dashboard/ActivityWidget.php [hunk 4]
```diff
@@ -72,17 +89,71 @@ public function getIconClass(): string {
...
+	 * @inheritDoc
+	 */
+	public function getWidgetButtons(string $userId): array {
+		return [
+			new WidgetButton(
+				WidgetButton::TYPE_MORE,
+				$this->urlGenerator->getAbsoluteURL(
+					$this->urlGenerator->linkToRoute(Application::APP_ID . '.Activities.showList')
+				),
+				$this->l10n->t('More Activities')
```
Comment (nickvergessen) [inline lib/Dashboard/ActivityWidget.php:155]:
```suggestion
				$this->l10n->t('More activities')
```

**Review by tobiasKaminsky — APPROVED** [review]
From client side :+1:

**Comment by nickvergessen** [comment]
Rebased for the psalm update

**Pull request merged**